x/group/server/invariants.go

```diff
@@ -0,0 +1,171 @@
+
+	var msg string
+	var broken bool
+
+	prevIt, err := proposalTable.PrefixScan(prevCtx, 1, math.MaxUint64)
+	if err != nil {
+		return msg, broken, err
+	}
+	curIt, err := proposalTable.PrefixScan(ctx, 1, math.MaxUint64)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	var curProposals []*group.Proposal
+	_, err = orm.ReadAll(curIt, &curProposals)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	var prevProposals []*group.Proposal
+	_, err = orm.ReadAll(prevIt, &prevProposals)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	for i := 0; i < len(prevProposals); i++ {
+		if prevProposals[i].ProposalId == curProposals[i].ProposalId {
+			prevYesCount, err := prevProposals[i].VoteState.GetYesCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			curYesCount, err := curProposals[i].VoteState.GetYesCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			prevNoCount, err := prevProposals[i].VoteState.GetNoCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			curNoCount, err := curProposals[i].VoteState.GetNoCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			prevAbstainCount, err := prevProposals[i].VoteState.GetAbstainCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			curAbstainCount, err := curProposals[i].VoteState.GetAbstainCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			prevVetoCount, err := prevProposals[i].VoteState.GetVetoCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			curVetoCount, err := curProposals[i].VoteState.GetVetoCount()
+			if err != nil {
+				return msg, broken, err
+			}
+			if (curYesCount.Cmp(prevYesCount) == -1) || (curNoCount.Cmp(prevNoCount) == -1) || (curAbstainCount.Cmp(prevAbstainCount) == -1) || (curVetoCount.Cmp(prevVetoCount) == -1) {
+				broken = true
+				msg += "vote tally sums must never have less than the block before\n"
+				return msg, broken, err
+			}
+		}
+	}
+	return msg, broken, err
+}
+
+func groupTotalWeightInvariant(ctx sdk.Context, groupTable orm.Table, groupMemberByGroupIndex orm.UInt64Index) (string, bool, error) {
+
+	var msg string
+	var broken bool
+
+	var groupInfo []*group.GroupInfo
+
+	groupIt, err := groupTable.PrefixScan(ctx, nil, nil)
+	if err != nil {
+		return msg, broken, err
+	}
+	_, err = orm.ReadAll(groupIt, &groupInfo)
+	if err != nil {
+		return msg, broken, err
+	}
+
+	var curGroupMembers []*group.GroupMember
+
+	membersWeight := apd.New(0, 0)
+
+	for i := 0; i < len(groupInfo); i++ {
+		memIt, err := groupMemberByGroupIndex.Get(ctx, groupInfo[i].GroupId)
+		if err != nil {
+			return msg, broken, err
+		}
+		_, err = orm.ReadAll(memIt, &curGroupMembers)
+		if err != nil {
+			return msg, broken, err
+		}
+
+		for j := 0; j < len(curGroupMembers); j++ {
+			curMemWeight, err := regenMath.ParseNonNegativeDecimal(curGroupMembers[j].GetMember().GetWeight())
```

Comment:
instead of using `ReadAll` then looping through all the `curGroupMembers` results, you could also directly use the `memIt` with `_, err := memIt.LoadNext(&member)` until there are no more items, when `ErrIteratorDone.Is(err)`
Also please for all the iterators within this file, don't forget to close them using `defer it.Close()` (there are lots of examples of this within the `orm` package)

Reply:
Okay